conn_js.go

```diff
@@ -0,0 +1,196 @@
+package websocket
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"syscall/js"
+	"time"
+)
+
+const (
+	webSocketStateConnecting = 0
+	webSocketStateOpen       = 1
+	webSocketStateClosing    = 2
+	webSocketStateClosed     = 3
+)
+
+var errIsClosed = errors.New("connection is closed")
+
+// Conn implements net.Conn interface for WebSockets in js/wasm.
+type Conn struct {
+	js.Value
+	messageHandler *js.Func
+	closeHandler   *js.Func
+	mut            sync.Mutex
+	incomingData   chan []byte
+	currData       *bytes.Buffer
+	closeSignal    chan struct{}
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	if err := c.checkOpen(); err != nil {
+		return 0, io.EOF
+	}
+
+	if c.currData == nil {
+		select {
+		case data := <-c.incomingData:
+			c.currData = bytes.NewBuffer(data)
+		case <-c.closeSignal:
+			return 0, io.EOF
+		}
+	}
+
+	n, err := c.currData.Read(b)
+	if err == io.EOF {
+		c.currData = nil
+		return n, nil
+	}
+	return n, err
+}
+
+func (c *Conn) checkOpen() error {
+	state := c.Get("readyState").Int()
+	switch state {
+	case webSocketStateClosed, webSocketStateClosing:
+		return errIsClosed
+	}
+	return nil
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	if err := c.checkOpen(); err != nil {
+		return 0, err
+	}
+	c.Call("send", string(b))
+	return len(b), nil
+}
+
+// Close closes the connection. Only the first call to Close will receive the
+// close error, subsequent and concurrent calls will return nil.
+// This method is thread-safe.
+func (c *Conn) Close() error {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	c.Call("close")
+	if c.messageHandler != nil {
+		c.Call("removeEventListener", "message", *c.messageHandler)
+		c.messageHandler.Release()
+	}
+	if c.closeHandler != nil {
+		c.Call("removeEventListener", "close", *c.closeHandler)
+		c.closeHandler.Release()
+	}
+	return nil
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	// TODO(albrow): is there some way to get the local address?
+	return NewAddr("0.0.0.0:0")
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	rawURL := c.Get("url").String()
+	withoutPrefix := strings.TrimPrefix(rawURL, "ws://")
+	withoutSuffix := strings.TrimSuffix(withoutPrefix, "/")
+	return NewAddr(withoutSuffix)
+}
+
+func (c *Conn) SetDeadline(t time.Time) error {
+	// Not supported
+	return nil
```

Comment:
These should return `os.ErrNoDeadline`.

Reply:
@Stebalien I tried making this change but it resulted in an error when trying to actually use the transport in a browser:

```
failed to negotiate security stream multiplexer: file type does not support deadline
```

I believe the error is coming from https://github.com/libp2p/go-libp2p-mplex or https://github.com/libp2p/go-libp2p-transport-upgrader.

Reply:
Ah. Can you leave a todo for now? We should both add support for this _and_ fix the multiplexer interfaces to take a context (we currently use deadlines to avoid this issue).

Reply:
👍 I added a TODO.